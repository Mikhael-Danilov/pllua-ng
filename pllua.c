/*
 * pllua.c: PL/Lua NG call handler
 * By Andrew "RhodiumToad" Gierth, rhodiumtoad at postgresql.org
 * Based in some part on pllua by Luis Carvalho and others
 * License: MIT license or PostgreSQL licence
 */

#include "pllua.h"

#include "commands/event_trigger.h"
#include "commands/trigger.h"

PG_MODULE_MAGIC;

/*
 * Exposed interface
 *
 * see also _PG_init in init.c
 */
PGDLLEXPORT Datum pllua_validator(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum pllua_call_handler(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum pllua_inline_handler(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum plluau_validator(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum plluau_call_handler(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum plluau_inline_handler(PG_FUNCTION_ARGS);

static Datum pllua_common_call(FunctionCallInfo fcinfo, bool trusted);
static Datum pllua_common_inline(FunctionCallInfo fcinfo, bool trusted);
static Datum pllua_common_validator(FunctionCallInfo fcinfo, bool trusted);


/* Trusted entry points */

PG_FUNCTION_INFO_V1(pllua_validator);
Datum pllua_validator(PG_FUNCTION_ARGS)
{
	return pllua_common_validator(fcinfo, true);
}

PG_FUNCTION_INFO_V1(pllua_call_handler);
Datum pllua_call_handler(PG_FUNCTION_ARGS)
{
	return pllua_common_call(fcinfo, true);
}

PG_FUNCTION_INFO_V1(pllua_inline_handler);
Datum pllua_inline_handler(PG_FUNCTION_ARGS)
{
	return pllua_common_inline(fcinfo, true);
}

/* Untrusted entry points */

PG_FUNCTION_INFO_V1(plluau_validator);
Datum plluau_validator(PG_FUNCTION_ARGS)
{
	return pllua_common_validator(fcinfo, false);
}

PG_FUNCTION_INFO_V1(plluau_call_handler);
Datum plluau_call_handler(PG_FUNCTION_ARGS)
{
	return pllua_common_call(fcinfo, false);
}

PG_FUNCTION_INFO_V1(plluau_inline_handler);
Datum plluau_inline_handler(PG_FUNCTION_ARGS)
{
	return pllua_common_inline(fcinfo, false);
}

/* Common implementations */

Datum pllua_common_call(FunctionCallInfo fcinfo, bool trusted)
{
	pllua_interpreter *interp;
	pllua_activation_record act;
	pllua_func_activation *funcact = (fcinfo->flinfo) ? fcinfo->flinfo->fn_extra : NULL;

	act.fcinfo = fcinfo;
	act.retval = (Datum) 0;
	act.trusted = trusted;
	act.cblock = NULL;
	act.validate_func = InvalidOid;

	pllua_setcontext(PLLUA_CONTEXT_PG);

	if (funcact && funcact->thread)
	{
		interp = funcact->interp;
		/*
		 * We're resuming a value-per-call SRF, so we bypass almost everything
		 * since we don't want to, for example, compile a new version of the
		 * function halfway through a result set. We know we're in a
		 * non-first-row condition if there's an existing thread in the
		 * function activation.
		 */
		pllua_initial_protected_call(interp, pllua_resume_function, &act);
	}
	else
	{
		interp = pllua_getstate(trusted, &act);

		if (CALLED_AS_TRIGGER(fcinfo))
			pllua_initial_protected_call(interp, pllua_call_trigger, &act);
		else if (CALLED_AS_EVENT_TRIGGER(fcinfo))
			pllua_initial_protected_call(interp, pllua_call_event_trigger, &act);
		else
			pllua_initial_protected_call(interp, pllua_call_function, &act);
	}

	return act.retval;
}

Datum pllua_common_validator(FunctionCallInfo fcinfo, bool trusted)
{
	pllua_interpreter *interp;
	pllua_activation_record act;
	Oid funcoid = PG_GETARG_OID(0);

	/* security checks */
	if (!CheckFunctionValidatorAccess(fcinfo->flinfo->fn_oid, funcoid))
		PG_RETURN_VOID();

	act.fcinfo = NULL;
	act.retval = (Datum) 0;
	act.trusted = trusted;
	act.cblock = NULL;
	act.validate_func = funcoid;

	pllua_setcontext(PLLUA_CONTEXT_PG);

	interp = pllua_getstate(trusted, &act);

	pllua_initial_protected_call(interp, pllua_validate, &act);

	PG_RETURN_VOID();
}

Datum pllua_common_inline(FunctionCallInfo fcinfo, bool trusted)
{
	pllua_interpreter *interp;
	pllua_activation_record act;

	/* XXX luajit may need this to be palloc'd. check later */

	act.fcinfo = NULL;
	act.retval = (Datum) 0;
	act.trusted = trusted;
	act.cblock = (InlineCodeBlock *) PG_GETARG_POINTER(0);
	act.validate_func = InvalidOid;

	pllua_setcontext(PLLUA_CONTEXT_PG);

	/* probably excess paranoia */
	if (act.cblock->langIsTrusted != act.trusted)
		elog(ERROR, "trusted state mismatch");

	interp = pllua_getstate(trusted, &act);

	pllua_initial_protected_call(interp, pllua_call_inline, &act);

	PG_RETURN_VOID();
}
