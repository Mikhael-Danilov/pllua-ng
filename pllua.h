/*
 * pllua.h
 */

#ifndef PLLUA_H
#define PLLUA_H

#include "postgres.h"

#include "fmgr.h"
#include "funcapi.h"

#include "utils/guc.h"
#include "utils/memutils.h"
#include "utils/palloc.h"

#include "miscadmin.h"

#include <lua.h>
#include <lualib.h>
#include <lauxlib.h>

#define PLLUA_LOCALVAR "_U"


#define pllua_pushcfunction(L_,f_) lua_pushcfunction((L_),(f_))

/*
 * Track what error handling context we're in, to try and detect any violations
 * of error-handling protocol (lua errors thrown through pg catch blocks and
 * vice-versa).
 */
typedef enum
{
	PLLUA_CONTEXT_PG,
	PLLUA_CONTEXT_LUA
} pllua_context_type;

extern pllua_context_type pllua_context;

#define ASSERT_PG_CONTEXT Assert(pllua_context == PLLUA_CONTEXT_PG)
#define ASSERT_LUA_CONTEXT Assert(pllua_context == PLLUA_CONTEXT_LUA)

static inline pllua_context_type
pllua_setcontext(pllua_context_type newctx)
{
	pllua_context_type oldctx = pllua_context;
	pllua_context = newctx;
	return oldctx;
}

/*
 * Abbreviate the most common form of catch block.
 */
#define PLLUA_TRY() do {												\
	pllua_context_type _pllua_oldctx = pllua_setcontext(PLLUA_CONTEXT_PG); \
	MemoryContext _pllua_oldmcxt = CurrentMemoryContext;				\
	PG_TRY()

#define PLLUA_CATCH_RETHROW()											\
	PG_CATCH();															\
	{ pllua_setcontext(_pllua_oldctx); pllua_rethrow_from_pg(L, _pllua_oldmcxt); } \
	PG_END_TRY(); pllua_setcontext(_pllua_oldctx); } while(0)

#define pllua_debug(L_, ...)											\
	do { if (pllua_context==PLLUA_CONTEXT_PG) elog(DEBUG1, __VA_ARGS__); \
		else pllua_debug_lua(L_, __VA_ARGS__); } while(0)

/*
 * We don't put this in the body of a lua userdata for error handling reasons;
 * we want to build it from pg data without involving lua too much until we're
 * ready to actually compile the function. Instead, the lua object is a pointer
 * to this with a __gc method, and the object itself is palloc'd. The
 * activation records (corresponding to flinfo) are lua objects that reference
 * the funcinfo, preventing it from being GC'd while in use.
 *
 * The actual lua function object is stored in the uservalue slot, and the
 * funcinfo has a __call method that proxies the call to it, so this object can
 * be treated as the function.
 */

typedef struct pllua_function_info
{
	Oid	fn_oid;
	TransactionId fn_xmin;
	ItemPointerData fn_tid;

	Oid rettype;
	bool retset;
	bool readonly;

	int nargs;
	bool variadic;
	bool variadic_any;
	bool polymorphic;

	Oid *argtypes;

	Oid language_oid;
	bool trusted;

	MemoryContext mcxt;

	const char *name;

} pllua_function_info;

typedef struct pllua_function_compile_info
{
	pllua_function_info *func_info;

	MemoryContext mcxt;

	text *prosrc;

	int nargs;
	int nallargs;

	Oid variadic;

	Oid *allargtypes;
	char *argmodes;
	char **argnames;

} pllua_function_compile_info;


/* this one ends up in flinfo->fn_extra */

typedef struct pllua_func_activation
{
	FmgrInfo *flinfo;
	pllua_function_info *func_info;
	lua_State *thread;

	bool resolved;

	bool polymorphic;
	bool variadic_call;
	bool retset;
	bool readonly;

	Oid rettype;
	TupleDesc tupdesc;
	TypeFuncClass typefuncclass;

	int nargs;
	Oid *argtypes;

	/*
	 * this data is allocated in lua, so we need to arrange to drop it for GC
	 * when the context containing the pointer to it is reset
	 */
	lua_State *L;
	MemoryContextCallback cb;
	bool dead;

} pllua_func_activation;


/*
 * Describes one call to the top-level handler.
 */
typedef struct pllua_activation_record
{
	FunctionCallInfo fcinfo;
	Datum		retval;
	bool		trusted;

	/* if fcinfo is null, we're validating or doing inline */
	InlineCodeBlock *cblock;
	Oid			validate_func;

} pllua_activation_record;

typedef struct pllua_datum {
	Datum value;
	int32 typmod;
	bool need_gc;
} pllua_datum;

typedef struct pllua_typeinfo {

	Oid typeoid;
	int32 typmod;  /* only for RECORD */

	/* 1 for scalars, otherwise natts - num_dropped_cols */
	int arity;

	/* -1 for scalars, since composites might have no columns */
	int natts;
	bool hasoid;
	bool revalidate;
	TupleDesc tupdesc;
	Oid reloid;  /* for named composite types */
	Oid basetype;  /* for domains */

	int16 typlen;
	bool typbyval;
	char typalign;
	char typdelim;
	Oid typioparam;
	Oid outfuncid;

	/* we don't look these up until we need them */
	Oid infuncid;
	Oid sendfuncid;
	Oid recvfuncid;

	FmgrInfo outfunc;
	FmgrInfo infunc;
	FmgrInfo sendfunc;
	FmgrInfo recvfunc;

	/* typmod coercions */
	bool coerce_typmod;
	bool coerce_typmod_element;
	Oid typmod_funcid;
	FmgrInfo typmod_func;

	/*
	 * we give this its own context, because we can't control what fmgr will
	 * dangle off the FmgrInfo structs
     */
	MemoryContext mcxt;

} pllua_typeinfo;

/*
 * are we shutting down?
 */

extern bool pllua_ending;

/*
 * Addresses used as lua registry or object keys
 *
 * Note the key is the address, not the string; the string is only for
 * debugging purposes.
 *
 * The registry looks like this (all keys are light userdata):
 *
 * global state:
 * reg[PLLUA_MEMORYCONTEXT] = light(MemoryContext) - for refobj data
 * reg[PLLUA_ERRORCONTEXT] = light(MemoryContext) - for error handling
 * reg[PLLUA_USERID] = int user_id for trusted, InvalidOid for untrusted
 * reg[PLLUA_TRUSTED] = boolean
 * reg[PLLUA_LAST_ERROR] = last pg error object to enter error handling
 * reg[PLLUA_RECURSIVE_ERROR] = preallocated error object
 *
 * registries for cached data:
 * reg[PLLUA_FUNCS] = { [integer oid] = funcinfo object }
 * reg[PLLUA_ACTIVATIONS] = { [light(act)] = activation object }
 * reg[PLLUA_TYPES] = { [integer oid] = typeinfo object }
 * reg[PLLUA_RECORDS] = { [integer typmod] = typeinfo object }
 *
 * metatables:
 * reg[PLLUA_FUNCTION_OBJECT]
 * reg[PLLUA_ACTIVATION_OBJECT]
 * reg[PLLUA_ERROR_OBJECT]
 * reg[PLLUA_TYPEINFO_OBJECT]
 * reg[PLLUA_TYPEINFO_PACKAGE_OBJECT]  (the pgtype() object itself)
 *  - datum objects and tupconv objects have dynamic metatables
 *
 * reg[PLLUA_]
 *
 *
 */

extern char PLLUA_MEMORYCONTEXT[];
extern char PLLUA_ERRORCONTEXT[];
extern char PLLUA_USERID[];
extern char PLLUA_TRUSTED[];
extern char PLLUA_FUNCS[];
extern char PLLUA_TYPES[];
extern char PLLUA_RECORDS[];
extern char PLLUA_ACTIVATIONS[];
extern char PLLUA_FUNCTION_OBJECT[];
extern char PLLUA_ERROR_OBJECT[];
extern char PLLUA_ACTIVATION_OBJECT[];
extern char PLLUA_TYPEINFO_OBJECT[];
extern char PLLUA_TYPEINFO_PACKAGE_OBJECT[];
extern char PLLUA_TUPCONV_OBJECT[];
extern char PLLUA_LAST_ERROR[];
extern char PLLUA_RECURSIVE_ERROR[];
extern char PLLUA_FUNCTION_MEMBER[];
extern char PLLUA_THREAD_MEMBER[];
extern char PLLUA_TYPEINFO_MEMBER[];

/* functions */

/* init.c */

lua_State *pllua_getstate(bool trusted);

/* compile.c */

pllua_func_activation *pllua_validate_and_push(lua_State *L, FunctionCallInfo fcinfo, bool trusted);

/* datum.c */

pllua_datum *pllua_checkanydatum(lua_State *L, int nd, pllua_typeinfo **ti);
void pllua_init_datum_objects(lua_State *L);
int pllua_typeinfo_invalidate(lua_State *L);
struct pllua_datum;
struct pllua_typeinfo;
void pllua_savedatum(lua_State *L,
					 struct pllua_datum *d,
					 struct pllua_typeinfo *t);
int pllua_value_from_datum(lua_State *L,
						   Datum value,
						   Oid typeid);
bool pllua_datum_from_value(lua_State *L, int nd,
							Oid typeid,
							Datum *result,
							bool *isnull);
pllua_datum *pllua_newdatum(lua_State *L);
void pllua_savedatum(lua_State *L,
					 pllua_datum *d,
					 pllua_typeinfo *t);
int pllua_typeinfo_lookup(lua_State *L);

/* elog.c */

int pllua_p_print (lua_State *L);
void pllua_init_error_functions(lua_State *L);
void pllua_debug_lua(lua_State *L, const char *msg, ...);

/* error.c */
int pllua_panic(lua_State *L);
void pllua_poperror(lua_State *L);
int pllua_newerror(lua_State *L);
int pllua_pcall_nothrow(lua_State *L, int nargs, int nresults, int msgh);
void pllua_rethrow_from_lua(lua_State *L, int rc);
void pllua_rethrow_from_pg(lua_State *L, MemoryContext mcxt);
int pllua_cpcall(lua_State *L, lua_CFunction func, void* arg);
void pllua_pcall(lua_State *L, int nargs, int nresults, int msgh);

void pllua_initial_protected_call(lua_State *L,
								  lua_CFunction func,
								  void *arg);

void pllua_init_error(lua_State *L);

/* exec.c */

int pllua_resume_function(lua_State *L);
int pllua_call_function(lua_State *L);
int pllua_call_trigger(lua_State *L);
int pllua_call_event_trigger(lua_State *L);
int pllua_call_inline(lua_State *L);
int pllua_validate(lua_State *L);

/* objects.c */

bool pllua_isobject(lua_State *L, int nd, char *objtype);
void pllua_newmetatable(lua_State *L, char *objtype, luaL_Reg *mt);
MemoryContext pllua_get_memory_cxt(lua_State *L);
void **pllua_newrefobject(lua_State *L, char *objtype, void *value, bool uservalue);
void **pllua_torefobject(lua_State *L, int nd, char *objtype);
void *pllua_newobject(lua_State *L, char *objtype, size_t sz, bool uservalue);
void *pllua_toobject(lua_State *L, int nd, char *objtype);
void pllua_type_error(lua_State *L, char *expected);
void **pllua_checkrefobject(lua_State *L, int nd, char *objtype);
void *pllua_checkobject(lua_State *L, int nd, char *objtype);

int pllua_newactivation(lua_State *L);
int pllua_setactivation(lua_State *L);
void pllua_getactivation(lua_State *L, pllua_func_activation *act);
int pllua_activation_getfunc(lua_State *L);
int pllua_get_cur_act(lua_State *L);
FmgrInfo *pllua_get_cur_flinfo(lua_State *L);

lua_State *pllua_activate_thread(lua_State *L, int nd, ExprContext *econtext);
void pllua_deactivate_thread(lua_State *L, pllua_func_activation *act, ExprContext *econtext);

void pllua_init_objects(lua_State *L, bool trusted);
void pllua_init_functions(lua_State *L, bool trusted);

#endif
