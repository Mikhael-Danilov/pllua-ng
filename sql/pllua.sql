--

CREATE EXTENSION pllua_ng;

\set VERBOSITY terse

-- smoke test

do language pllua_ng $$ print "hello world!" $$;
do language plluau_ng $$ print "hello world!" $$;

create function pg_temp.f1() returns text language pllua_ng as $$ return "hello world" $$;
select pg_temp.f1();

create function pg_temp.f2() returns text language plluau_ng as $$ return "hello world" $$;
select pg_temp.f2();

-- Rest of this file concentrates on simple tests of code paths in
-- compile, exec, and interpreter setup. Tests of other parts of the
-- module are separate.

-- validator
create function pg_temp."bad name"() returns text language pllua_ng as $$ $$;
create function pg_temp.f3("bad arg" integer) returns text language pllua_ng as $$ $$;

-- simple params and results (see types.sql for detailed checks)

create function pg_temp.f4(a integer) returns integer language pllua_ng as $$ return a + 1 $$;
select pg_temp.f4(1);

create function pg_temp.f5(a text) returns text language pllua_ng as $$ return a.."bar" $$;
select pg_temp.f5('foo');

create function pg_temp.f6(a text, b integer) returns text language pllua_ng as $$ return a..b $$;
select pg_temp.f6('foo',1);

-- try some polymorphism too
create function pg_temp.f7(a anyelement) returns anyelement language pllua_ng as $$ return a $$;
select pg_temp.f7(text 'foo');
select pg_temp.f7(json '{"foo":1}');
select pg_temp.f7(xml '<foo>bar</foo>');
select pg_temp.f7(varchar 'foo');
select 'x',pg_temp.f7('foo'::char(20)),'x';
select pg_temp.f7(cstring 'foo');
select pg_temp.f7(name 'foo');
select pg_temp.f7(bytea 'foo\000bar');
select pg_temp.f7(smallint '2');
select pg_temp.f7(integer '2');
select pg_temp.f7(bigint '123456789012345');
select pg_temp.f7(oid '10');
select pg_temp.f7(oid '4294967295');
select pg_temp.f7(true);
select pg_temp.f7(false);
select pg_temp.f7(1.5::float8);
select pg_temp.f7(1.5::float4);

-- variadics

create function pg_temp.f8(a text, variadic b integer[]) returns void language pllua_ng as $$ print(a,type(b),b) $$;
select pg_temp.f8('foo', 1, 2, 3);

create function pg_temp.f9(a integer, variadic b text[]) returns void language pllua_ng as $$ print(a,type(b),b) $$;
select pg_temp.f9(1, 'foo', 'bar', 'baz');

create function pg_temp.f10(a integer, variadic "any") returns void language pllua_ng as $$ print(a,...) $$;
select pg_temp.f10(1, 'foo', 2, 'baz');

-- SRF code paths

create function pg_temp.f11(a integer) returns setof text
  language pllua_ng as $$ return $$;  -- 0 rows
select * from pg_temp.f11(1);

create function pg_temp.f11b(a integer) returns setof text
  language pllua_ng as $$ return 'foo' $$;  -- 1 row
select * from pg_temp.f11b(1);

create function pg_temp.f12(a integer) returns setof text
  language pllua_ng as $$ coroutine.yield() $$;  -- 1 row, null
select * from pg_temp.f12(1);

create function pg_temp.f13(a integer) returns setof text
  language pllua_ng as $$ for i = 1,a do coroutine.yield("row "..i) end $$;
select * from pg_temp.f13(4);

create function pg_temp.f14(a integer, out x text, out y integer) returns setof record
  language pllua_ng as $$ for i = 1,a do coroutine.yield("row "..i, i) end $$;
select * from pg_temp.f14(4);

create function pg_temp.f15(a integer) returns table(x text, y integer)
  language pllua_ng as $$ for i = 1,a do coroutine.yield("row "..i, i) end $$;
select * from pg_temp.f15(4);

create function pg_temp.f16(a inout integer, x out text) returns setof record
  language pllua_ng as $$ for i = 1,a do coroutine.yield(i, "row "..i) end $$;
select * from pg_temp.f16(4);

-- compiler and validator code paths

do language pllua_ng $$ rdepth = 40 $$;  -- global var hack

-- This function will try and call itself at a point where it is visible
-- but has no definition interned yet; the recursive call will likewise
-- not see an interned definition and recurses again. without any limits
-- this would hit a stack depth check somewhere; we eat about 3 levels of
-- C function recursion inside lua each time, and that gets capped at 200.
-- We don't expect this to be actually useful, the test is just that we
-- don't crash.
create function pg_temp.f17(a integer) returns integer language pllua_ng
  as $$
    return a
  end
  do
    if rdepth > 0 then
      rdepth = rdepth - 1
      _U = spi.execute("select pg_temp.f17(1)")
    end
$$;
select pg_temp.f17(1);

create type pg_temp.t1 as (a integer, b text);
create function pg_temp.f18(a integer, b text) returns pg_temp.t1
  language pllua_ng as $$ return a,b $$;
select * from pg_temp.f18(123,'foo');

create function pg_temp.f19(a integer) returns text language pllua_ng as $$ return 'foo '..a $$;
select pg_temp.f19(2);
create or replace function pg_temp.f19(a integer) returns text language pllua_ng as $$ return 'bar '..a $$;
select pg_temp.f19(3);

-- trusted interpreter setup

-- check we really do have different interpreters
-- (uses table.x rather than _ENV.x because _ENV would be sandboxed)
do language pllua_ng $$ table.g_trusted = 1 $$;
do language plluau_ng $$ table.g_untrusted = 1 print('untrusted',table.g_trusted,table.g_untrusted)$$;
do language pllua_ng $$ print('trusted',table.g_trusted,table.g_untrusted) $$;

-- check the global table
do language pllua_ng $$
  local gk = { "io", "dofile", "debug" }  -- must not exist
  for i = 1,#gk do print(gk[i],type(_ENV[gk[i]])) end
$$;
do language plluau_ng $$
  local gk = { "io", "dofile" }  -- probably exist
  for i = 1,#gk do print(gk[i],type(_ENV[gk[i]])) end
$$;

--end
