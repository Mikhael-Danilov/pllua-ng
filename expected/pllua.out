--
CREATE EXTENSION pllua_ng;
\set VERBOSITY terse
-- smoke test
do language pllua_ng $$ print "hello world!" $$;
INFO:  hello world!
do language plluau_ng $$ print "hello world!" $$;
INFO:  hello world!
create function pg_temp.f1() returns text language pllua_ng as $$ return "hello world" $$;
select pg_temp.f1();
     f1      
-------------
 hello world
(1 row)

create function pg_temp.f2() returns text language plluau_ng as $$ return "hello world" $$;
select pg_temp.f2();
     f2      
-------------
 hello world
(1 row)

-- Rest of this file concentrates on simple tests of code paths in
-- compile, exec, and interpreter setup. Tests of other parts of the
-- module are separate.
-- validator
create function pg_temp."bad name"() returns text language pllua_ng as $$ $$;
ERROR:  PL/Lua function name "bad name" is not a valud Lua identifier
create function pg_temp.f3("bad arg" integer) returns text language pllua_ng as $$ $$;
ERROR:  PL/Lua argument name "bad arg" is not a valud Lua identifier
-- simple params and results (see types.sql for detailed checks)
create function pg_temp.f4(a integer) returns integer language pllua_ng as $$ return a + 1 $$;
select pg_temp.f4(1);
 f4 
----
  2
(1 row)

create function pg_temp.f5(a text) returns text language pllua_ng as $$ return a.."bar" $$;
select pg_temp.f5('foo');
   f5   
--------
 foobar
(1 row)

create function pg_temp.f6(a text, b integer) returns text language pllua_ng as $$ return a..b $$;
select pg_temp.f6('foo',1);
  f6  
------
 foo1
(1 row)

-- try some polymorphism too
create function pg_temp.f7(a anyelement) returns anyelement language pllua_ng as $$ return a $$;
select pg_temp.f7(text 'foo');
 f7  
-----
 foo
(1 row)

select pg_temp.f7(json '{"foo":1}');
    f7     
-----------
 {"foo":1}
(1 row)

select pg_temp.f7(xml '<foo>bar</foo>');
       f7       
----------------
 <foo>bar</foo>
(1 row)

select pg_temp.f7(varchar 'foo');
 f7  
-----
 foo
(1 row)

select 'x',pg_temp.f7('foo'::char(20)),'x';
 ?column? |          f7          | ?column? 
----------+----------------------+----------
 x        | foo                  | x
(1 row)

select pg_temp.f7(cstring 'foo');
 f7  
-----
 foo
(1 row)

select pg_temp.f7(name 'foo');
 f7  
-----
 foo
(1 row)

select pg_temp.f7(bytea 'foo\000bar');
        f7        
------------------
 \x666f6f00626172
(1 row)

select pg_temp.f7(smallint '2');
 f7 
----
  2
(1 row)

select pg_temp.f7(integer '2');
 f7 
----
  2
(1 row)

select pg_temp.f7(bigint '123456789012345');
       f7        
-----------------
 123456789012345
(1 row)

select pg_temp.f7(oid '10');
 f7 
----
 10
(1 row)

select pg_temp.f7(oid '4294967295');
     f7     
------------
 4294967295
(1 row)

select pg_temp.f7(true);
 f7 
----
 t
(1 row)

select pg_temp.f7(false);
 f7 
----
 f
(1 row)

select pg_temp.f7(1.5::float8);
 f7  
-----
 1.5
(1 row)

select pg_temp.f7(1.5::float4);
 f7  
-----
 1.5
(1 row)

-- variadics
create function pg_temp.f8(a text, variadic b integer[]) returns void language pllua_ng as $$ print(a,type(b),b) $$;
select pg_temp.f8('foo', 1, 2, 3);
INFO:  foo	userdata	{1,2,3}
 f8 
----
 
(1 row)

create function pg_temp.f9(a integer, variadic b text[]) returns void language pllua_ng as $$ print(a,type(b),b) $$;
select pg_temp.f9(1, 'foo', 'bar', 'baz');
INFO:  1	userdata	{foo,bar,baz}
 f9 
----
 
(1 row)

create function pg_temp.f10(a integer, variadic "any") returns void language pllua_ng as $$ print(a,...) $$;
select pg_temp.f10(1, 'foo', 2, 'baz');
INFO:  1	foo	2	baz
 f10 
-----
 
(1 row)

-- SRF code paths
create function pg_temp.f11(a integer) returns setof text
  language pllua_ng as $$ return $$;  -- 0 rows
select * from pg_temp.f11(1);
 f11 
-----
(0 rows)

create function pg_temp.f11b(a integer) returns setof text
  language pllua_ng as $$ return 'foo' $$;  -- 1 row
select * from pg_temp.f11b(1);
 f11b 
------
 foo
(1 row)

create function pg_temp.f12(a integer) returns setof text
  language pllua_ng as $$ coroutine.yield() $$;  -- 1 row, null
select * from pg_temp.f12(1);
 f12 
-----
 
(1 row)

create function pg_temp.f13(a integer) returns setof text
  language pllua_ng as $$ for i = 1,a do coroutine.yield("row "..i) end $$;
select * from pg_temp.f13(4);
  f13  
-------
 row 1
 row 2
 row 3
 row 4
(4 rows)

create function pg_temp.f14(a integer, out x text, out y integer) returns setof record
  language pllua_ng as $$ for i = 1,a do coroutine.yield("row "..i, i) end $$;
select * from pg_temp.f14(4);
   x   | y 
-------+---
 row 1 | 1
 row 2 | 2
 row 3 | 3
 row 4 | 4
(4 rows)

create function pg_temp.f15(a integer) returns table(x text, y integer)
  language pllua_ng as $$ for i = 1,a do coroutine.yield("row "..i, i) end $$;
select * from pg_temp.f15(4);
   x   | y 
-------+---
 row 1 | 1
 row 2 | 2
 row 3 | 3
 row 4 | 4
(4 rows)

create function pg_temp.f16(a inout integer, x out text) returns setof record
  language pllua_ng as $$ for i = 1,a do coroutine.yield(i, "row "..i) end $$;
select * from pg_temp.f16(4);
 a |   x   
---+-------
 1 | row 1
 2 | row 2
 3 | row 3
 4 | row 4
(4 rows)

-- compiler and validator code paths
do language pllua_ng $$ rdepth = 40 $$;  -- global var hack
-- This function will try and call itself at a point where it is visible
-- but has no definition interned yet; the recursive call will likewise
-- not see an interned definition and recurses again. without any limits
-- this would hit a stack depth check somewhere; we eat about 3 levels of
-- C function recursion inside lua each time, and that gets capped at 200.
-- We don't expect this to be actually useful, the test is just that we
-- don't crash.
create function pg_temp.f17(a integer) returns integer language pllua_ng
  as $$
    return a
  end
  do
    if rdepth > 0 then
      rdepth = rdepth - 1
      _U = spi.execute("select pg_temp.f17(1)")
    end
$$;
select pg_temp.f17(1);
 f17 
-----
   1
(1 row)

create type pg_temp.t1 as (a integer, b text);
create function pg_temp.f18(a integer, b text) returns pg_temp.t1
  language pllua_ng as $$ return a,b $$;
select * from pg_temp.f18(123,'foo');
  a  |  b  
-----+-----
 123 | foo
(1 row)

create function pg_temp.f19(a integer) returns text language pllua_ng as $$ return 'foo '..a $$;
select pg_temp.f19(2);
  f19  
-------
 foo 2
(1 row)

create or replace function pg_temp.f19(a integer) returns text language pllua_ng as $$ return 'bar '..a $$;
select pg_temp.f19(3);
  f19  
-------
 bar 3
(1 row)

-- trusted interpreter setup
-- check we really do have different interpreters
-- (uses table.x rather than _ENV.x because _ENV would be sandboxed)
do language pllua_ng $$ table.g_trusted = 1 $$;
do language plluau_ng $$ table.g_untrusted = 1 print('untrusted',table.g_trusted,table.g_untrusted)$$;
INFO:  untrusted	nil	1
do language pllua_ng $$ print('trusted',table.g_trusted,table.g_untrusted) $$;
INFO:  trusted	1	nil
-- check the global table
do language pllua_ng $$
  local gk = { "io", "dofile", "debug" }  -- must not exist
  for i = 1,#gk do print(gk[i],type(_ENV[gk[i]])) end
$$;
INFO:  io	nil
INFO:  dofile	nil
INFO:  debug	nil
do language plluau_ng $$
  local gk = { "io", "dofile" }  -- probably exist
  for i = 1,#gk do print(gk[i],type(_ENV[gk[i]])) end
$$;
INFO:  io	table
INFO:  dofile	function
--end
